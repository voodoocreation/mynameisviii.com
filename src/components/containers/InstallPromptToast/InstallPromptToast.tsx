import moment from "moment";
import * as React from "react";
import { FormattedMessage, InjectedIntl, injectIntl } from "react-intl";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { ActionCreator } from "typescript-fsa";

import Button from "../../presentation/Button/Button";
import Toast from "../../presentation/Toast/Toast";

import * as actions from "../../../actions/root.actions";
import {
  getLocalStorage,
  setLocalStorage
} from "../../../services/configureLocalStorage";

interface IDispatchProps {
  trackEvent: ActionCreator<PLTrackEvent>;
}

interface IProps extends IDispatchProps {
  intl: InjectedIntl;
}

interface IState {
  isVisible: boolean;
}

class InstallPromptToast extends React.Component<IProps, IState> {
  public readonly state = {
    isVisible: false
  };

  private deferredEvent: any = undefined;

  public componentDidMount() {
    window.addEventListener("beforeinstallprompt", this.onBeforeInstallPrompt);
  }

  public componentWillUnmount() {
    window.removeEventListener(
      "beforeinstallprompt",
      this.onBeforeInstallPrompt
    );
  }

  public render() {
    return this.state.isVisible ? (
      <Toast hasAutoDismiss={false} onClose={this.onClose}>
        <FormattedMessage id="ADD_TO_HOME_SCREEN" />
        <Button className="Toast-installButton" onClick={this.onInstallClick}>
          <FormattedMessage id="INSTALL" />
        </Button>
      </Toast>
    ) : null;
  }

  private onClose = async () => {
    this.props.trackEvent({
      event: "addToHomeScreen",
      outcome: "toast closed"
    });

    setLocalStorage("addToHomeScreen.dismissed", `${Date.now()}`);
  };

  private onBeforeInstallPrompt = (event: any) => {
    event.preventDefault();
    this.deferredEvent = event;

    const { data } = getLocalStorage("addToHomeScreen.dismissed");
    const dismissedAt = moment.unix(Number(data));
    const expiredAt = moment(new Date()).subtract(3, "days");

    if (!data || expiredAt.isSameOrAfter(dismissedAt)) {
      this.setState({
        isVisible: true
      });
    }
  };

  private onInstallClick = async () => {
    this.setState({
      isVisible: false
    });

    this.deferredEvent.prompt();
    const { outcome } = await this.deferredEvent.userChoice;

    this.props.trackEvent({
      event: "addToHomeScreen",
      outcome
    });

    if (outcome === "dismissed") {
      setLocalStorage("addToHomeScreen.dismissed", `${Date.now()}`);
    }

    this.deferredEvent = undefined;
  };
}

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      trackEvent: actions.trackEvent
    },
    dispatch
  );

export default injectIntl<any>(
  connect<undefined, IDispatchProps>(
    undefined,
    mapDispatchToProps
  )(InstallPromptToast)
);
