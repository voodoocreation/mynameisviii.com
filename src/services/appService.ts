const g: any = global;
const s: any = self;
const { assets, buildId, staticFiles } = g.serviceWorkerOption;

g.CACHE_PREFIX = "mynameisviii.com";
g.cacheNames = {
  api: `${g.CACHE_PREFIX}-api`,
  cdn: `${g.CACHE_PREFIX}-cdn`,
  googleApis: `${g.CACHE_PREFIX}-googleApis`,
  local: `${g.CACHE_PREFIX}-local-${buildId}`,
  precache: `${g.CACHE_PREFIX}-precache-${buildId}`
};

g.precacheUrls = [
  ...assets,
  ...staticFiles
];

const isCacheValid = (key: string) => Object.values(g.cacheNames).includes(key);

g.importScripts(
  "https://storage.googleapis.com/workbox-cdn/releases/3.4.1/workbox-sw.js"
);

g.workbox.core.setCacheNameDetails({
  prefix: g.CACHE_PREFIX,
  suffix: buildId
});

g.workbox.skipWaiting();
g.workbox.clientsClaim();

if (buildId !== "development") {
  g.workbox.precaching.precacheAndRoute(
    g.precacheUrls.map((item: string) => ({
      revision: buildId,
      url: item
    })),
    {
      cleanUrls: false
    }
  );

  // Local cache-first requests
  g.workbox.routing.registerRoute(
    ({ url }: any) =>
      url.host === self.location.host && !g.precacheUrls.includes(url.pathname),
    g.workbox.strategies.cacheFirst({
      cacheName: g.cacheNames.local,
      plugins: [
        new g.workbox.cacheableResponse.Plugin({
          statuses: [0, 200]
        }),
        new g.workbox.expiration.Plugin({
          maxAgeSeconds: 2 * 60 * 60
        })
      ]
    })
  );
}

g.workbox.googleAnalytics.initialize({
  cacheName: g.cacheNames.googleApis
});

// Static content requests from Google APIs
g.workbox.routing.registerRoute(
  new RegExp("https://(?:fonts).(?:googleapis|gstatic).com/(.*)"),
  g.workbox.strategies.cacheFirst({
    cacheName: g.cacheNames.googleApis,
    plugins: [
      new g.workbox.cacheableResponse.Plugin({
        statuses: [0, 200]
      })
    ]
  })
);

// Dynamic content requests from Google APIs
g.workbox.routing.registerRoute(
  new RegExp("https://(?:maps).(?:googleapis|gstatic).com/(.*)"),
  g.workbox.strategies.networkOnly({
    cacheName: g.cacheNames.googleApis,
    plugins: [
      new g.workbox.cacheableResponse.Plugin({
        statuses: [0, 200]
      })
    ]
  })
);

// App CDN requests
g.workbox.routing.registerRoute(
  new RegExp("https://s3.amazonaws.com/mynameisviii-static/(.*)"),
  g.workbox.strategies.cacheFirst({
    cacheName: g.cacheNames.cdn,
    plugins: [
      new g.workbox.cacheableResponse.Plugin({
        statuses: [0, 200]
      })
    ]
  })
);

// App API requests
// Listings - cache for 2 hours
g.workbox.routing.registerRoute(
  new RegExp("https://api.mynameisviii.com/(.*)/find(.*)"),
  g.workbox.strategies.cacheFirst({
    cacheName: g.cacheNames.api,
    plugins: [
      new g.workbox.cacheableResponse.Plugin({
        statuses: [0, 200]
      }),
      new g.workbox.expiration.Plugin({
        maxAgeSeconds: 2 * 60 * 60
      })
    ]
  })
);
// Entities - cache for a week
g.workbox.routing.registerRoute(
  new RegExp("https://api.mynameisviii.com/(.*)/[^find](.*)"),
  g.workbox.strategies.cacheFirst({
    cacheName: g.cacheNames.api,
    plugins: [
      new g.workbox.cacheableResponse.Plugin({
        statuses: [0, 200]
      }),
      new g.workbox.expiration.Plugin({
        maxAgeSeconds: 7 * 24 * 60 * 60
      })
    ]
  })
);

self.addEventListener("activate", async (event: any) => {
  // Cleanup redundant caches
  event.waitUntil(
    caches.keys().then(async keys => {
      let hasDeletedCaches = false;
      keys.forEach(key => {
        if (!isCacheValid(key)) {
          caches.delete(key);
          indexedDB.deleteDatabase(key);
          hasDeletedCaches = true;
        }
      });

      // Notify client that a new version is available, when an old version exists
      if (hasDeletedCaches) {
        const clients = await s.clients.matchAll();
        clients.forEach((client: any) => {
          client.postMessage({
            type: "serviceWorker.activate"
          });
        });
      }
    })
  );
});

// Handle postMessage requests from the application
self.onmessage = message => {
  switch (message.data.type) {
    case "changeRoute":
      return caches
        .open(g.cacheNames.local)
        .then(cache => cache.add(message.data.payload));
  }
};
