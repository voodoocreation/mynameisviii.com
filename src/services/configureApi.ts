import axios from "axios";
import { camelizeKeys } from "humps";

import * as api from "./api/root.api";

const mapApi = (methods: any, ports: any) =>
  Object.keys(methods).reduce(
    (acc: any, curr: any) => ({
      ...acc,
      [curr]: methods[curr](ports)
    }),
    {}
  );

export const createApiWith = (ports: any) => mapApi(api, ports);

export const createPortsWith = (
  config: any,
  client: (config: any) => Promise<any> = axios
) => ({
  body,
  method = "GET",
  params,
  url
}: {
  body?: string;
  method?: string;
  params?: {};
  url: string;
}) =>
  client({
    baseURL: config.apiUrl,
    data: body,
    method,
    params,
    url
  })
    .then(response => camelizeKeys(response.data))
    .catch(error => {
      if (error.response) {
        // The request was made and the server responded with a status code
        // that falls out of the range of 2xx
        throw new Error(
          JSON.stringify(
            camelizeKeys({
              ...error.response.data,
              status: error.response.status
            })
          )
        );
      } else if (error.request) {
        // The request was made but no response was received
        // `err.request` is an instance of XMLHttpRequest in the browser
        throw new Error(error.request.statusText);
      } else {
        // Something happened in setting up the request that triggered an Error
        throw new Error(error.message);
      }
    });
