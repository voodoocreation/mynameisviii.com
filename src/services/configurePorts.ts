import { configureApi, configureMockApi, TApi, TMockApi } from "./configureApi";
import { configureHttpClient } from "./configureHttpClient";
import { configureLocalStorage } from "./configureLocalStorage";

export interface IPorts {
  api: TApi;
  dataLayer: any[];
  features: string[];
  maps: {
    [index: string]: any;
  };
}

export interface IPortsConfig {
  dataLayer?: any[];
  features?: string[];
  fetch: typeof window.fetch;
  maps?: {
    [index: string]: any;
  };
}

export interface ITestPorts {
  api: TMockApi;
  dataLayer: any[];
  features: string[];
  maps: {
    [index: string]: any;
  };
}

export interface ITestPortsParam {
  api?: Partial<TMockApi>;
  dataLayer?: any[];
  features?: string[];
  maps?: {
    [index: string]: any;
  };
}

export const configurePorts = (config: IPortsConfig): IPorts => {
  const dataLayer: any[] = config.dataLayer || [];
  dataLayer.push = dataLayer.push.bind(dataLayer);

  const api = configureApi(
    configureHttpClient({
      fetch: config.fetch
    }),
    configureLocalStorage()
  );

  return {
    api,
    dataLayer,
    features: config.features || [],
    maps: config.maps || {}
  };
};

export const configureTestPorts = (ports: ITestPortsParam = {}): ITestPorts => {
  const dataLayer: any[] = ports.dataLayer ? ports.dataLayer : [];
  const api: TMockApi = { ...configureMockApi(), ...(ports.api as any) };

  return {
    api,
    dataLayer,
    features: ports.features || [],
    maps: {
      ...window.google.maps,
      ...(ports.maps || {})
    }
  };
};
