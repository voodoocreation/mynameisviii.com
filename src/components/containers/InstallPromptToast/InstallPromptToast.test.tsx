import { mount, render } from "enzyme";
import moment from "moment";
import * as React from "react";

import InstallPromptToast from "./InstallPromptToast";

import * as ls from "../../../services/configureLocalStorage";

jest.mock("react-redux", () => ({
  connect: () => (component: any) => component
}));

jest.mock("../../../services/configureLocalStorage", () => ({
  getLocalStorage: jest.fn(() => ({ data: null, ok: true })),
  setLocalStorage: jest.fn(() => ({ data: null, ok: true }))
}));

const setup = (fn: any, fromTestProps: any = {}) => {
  const props = {
    trackEvent: jest.fn(),
    ...fromTestProps
  };
  return {
    actual: fn(<InstallPromptToast {...props} />),
    props
  };
};

describe("[containers] <InstallPromptToast />", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders null initially", () => {
    const { actual } = setup(render);

    expect(actual).toMatchSnapshot();
  });

  it("displays correctly when `beforeinstallprompt` event is fired", () => {
    const { actual } = setup(mount);

    const event: any = new Event("beforeinstallprompt");
    window.dispatchEvent(event);
    actual.update();

    expect(actual.find("Toast")).toHaveLength(1);
    expect(actual.render()).toMatchSnapshot();
    actual.unmount();
  });

  it("displays when the user has previously dismissed the toast over three days ago", () => {
    const getLocalStorage: any = ls.getLocalStorage;
    getLocalStorage.mockImplementationOnce(() => ({
      data: `${moment(new Date())
        .subtract(4, "days")
        .unix()}`,
      ok: true
    }));

    const { actual } = setup(mount);

    const event: any = new Event("beforeinstallprompt");
    window.dispatchEvent(event);
    actual.update();

    expect(actual.find("Toast")).toHaveLength(1);
    actual.unmount();
  });

  it("doesn't display when the user has previously dismissed the toast less than three days ago", () => {
    const getLocalStorage: any = ls.getLocalStorage;
    getLocalStorage.mockImplementationOnce(() => ({
      data: `${moment(new Date())
        .subtract(2, "days")
        .unix()}`,
      ok: true
    }));

    const { actual } = setup(mount);

    const event: any = new Event("beforeinstallprompt");
    window.dispatchEvent(event);
    actual.update();

    expect(actual.find("Toast")).toHaveLength(0);
    actual.unmount();
  });

  it("tracks event and dismisses toast correctly when the close button is clicked", () => {
    const { actual, props } = setup(mount);
    const setLocalStorage: any = ls.setLocalStorage;

    const event: any = new Event("beforeinstallprompt");
    window.dispatchEvent(event);
    actual.update();

    actual.find(".Toast-closeButton").simulate("click");

    expect(props.trackEvent).toHaveBeenCalledWith({
      event: "addToHomeScreen",
      outcome: "toast closed"
    });
    expect(setLocalStorage).toHaveBeenCalled();
    expect(setLocalStorage.mock.calls[0][0]).toBe("addToHomeScreen.dismissed");
    actual.unmount();
  });

  describe("when the install button is clicked", () => {
    it("hides the toast and prompts the user to install the app, then dismisses correctly when the user cancels the prompt", async () => {
      const { actual, props } = setup(mount);
      const setLocalStorage: any = ls.setLocalStorage;

      const event: any = new Event("beforeinstallprompt");
      event.prompt = jest.fn();
      event.userChoice = new Promise(resolve =>
        resolve({ outcome: "dismissed" })
      );
      window.dispatchEvent(event);
      actual.update();

      expect(event.prompt).not.toHaveBeenCalled();

      await actual
        .find(".Toast-installButton")
        .first()
        .simulate("click");

      expect(event.prompt).toHaveBeenCalled();
      expect(props.trackEvent).toHaveBeenCalledWith({
        event: "addToHomeScreen",
        outcome: "dismissed"
      });
      expect(setLocalStorage).toHaveBeenCalled();
      expect(setLocalStorage.mock.calls[0][0]).toBe(
        "addToHomeScreen.dismissed"
      );
      actual.unmount();
    });

    it("hides the toast and prompts the user to install the app, then dismisses correctly when the user accepts the prompt", async () => {
      const { actual, props } = setup(mount);
      const setLocalStorage: any = ls.setLocalStorage;

      const event: any = new Event("beforeinstallprompt");
      event.prompt = jest.fn();
      event.userChoice = new Promise(resolve =>
        resolve({ outcome: "accepted" })
      );
      window.dispatchEvent(event);
      actual.update();

      expect(event.prompt).not.toHaveBeenCalled();

      await actual
        .find(".Toast-installButton")
        .first()
        .simulate("click");

      expect(event.prompt).toHaveBeenCalled();
      expect(setLocalStorage).not.toHaveBeenCalled();
      expect(props.trackEvent).toHaveBeenCalledWith({
        event: "addToHomeScreen",
        outcome: "accepted"
      });
      actual.unmount();
    });
  });
});
