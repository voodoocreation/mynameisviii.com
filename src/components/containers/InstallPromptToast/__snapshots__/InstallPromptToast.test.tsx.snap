// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[containers] <InstallPromptToast /> displays correctly when \`beforeinstallprompt\` event is fired 1`] = `
<div
  class="Toast isHidden"
>
  <div
    class="Toast-content"
  >
    <div
      class="Toast-message"
    >
      <span>
        Would you like to install the Viii app?
      </span>
      <button
        class="Button Toast-installButton"
        type="button"
      >
        <span
          class="Button-content"
        >
          <span>
            Install
          </span>
        </span>
      </button>
    </div>
    <button
      class="Toast-closeButton"
      title="Close"
    >
      <svg
        attr="[object Object]"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`[containers] <InstallPromptToast /> renders null initially 1`] = `null`;
